Increase performance when checking if element is animatable

Narretz opened the pull request:
Built upon PR #13783

Narretz committed:
fix($animate): correctly handle `$animate.pin()` host elements

This commit fixes two bugs:
1) Previously, animate would assume that a found host element
was part of the $rootElement (while it's possible that it is also outside the root).

2) Previously, if a parent of the animated element was pinned to a host element, the
host would not be checked regarding animations enabled status etc.

Closes #13783

gkalpak commented on test/ngAnimate/animateSpec.js:
```diff
@@ -1543,6 +1543,49 @@ describe("animations", function() {
+    they('should not animate an element when the pinned element (which is the $prop element), is pinned to an element that is not a child of the $rootElement',
+      ['same', 'parent', 'grandparent'],
+      function(elementRelation) {
+        inject(function($animate, $compile, $document, $rootElement, $rootScope) {
```
`$compile` is not used.
(Yes, in other tests too :smiley:)

gkalpak commented on test/ngAnimate/animateSpec.js:
```diff
@@ -1543,6 +1543,49 @@ describe("animations", function() {
+        $rootScope.$digest();
+        expect(capturedAnimation).toBeFalsy();
+
+        dealoc(pinElement);
```
Why do we only dealoc `pinElement` ?
What about `pinTargetElement` or `innerParent`.

(Other tests seem inconsistent too (unless I'm missing something.)

gkalpak commented on src/ngAnimate/animateQueue.js:
```diff
@@ -579,6 +586,11 @@ var $$AnimateQueueProvider = ['$animateProvider', function($animateProvider) {
+      if (elementDisabled === true) {
+        // If animations on the element have been disabled explicitly, no need to traverse the parents
+        return false;
+      }
```
AFAICT, `elementDisabled` could be still set to `false` on [line 626][1], so this change is either invalid or also a fix (in addition to perf improvement). I believe it is the latter (in which case a test should be added).

[1]: https://github.com/angular/angular.js/pull/13784/files#diff-c15d87c15661cba4a721501ba4fda5b2R626

Narretz replied:
But line 626 is inside the loop, and this check comes before the loop. Also, this is checked before areAnimationsAllowed, too. The lookup here was added later to catch elements that were explicitly allowed.

Narretz replied:
I'll remove this, because this is checked already at https://github.com/angular/angular.js/pull/13784/files#diff-c15d87c15661cba4a721501ba4fda5b2R342

gkalpak commented on test/ngAnimate/animateSpec.js:
```diff
@@ -1543,6 +1543,49 @@ describe("animations", function() {
       });
     });
 
+    they('should not animate an element when the pinned element (which is the $prop element), is pinned to an element that is not a child of the $rootElement',
```
This line is too long :smiley:"

Narretz replied:
Really? Shouldn't jshint / jscs report this then? I thought we were at 200 chars.

gkalpak replied:
Wasn't it at 100 chars ? Did that change and I missed that ?
Imo, 100 chars makes sense because it makes side-by-side diffing much easier.

gkalpak commented on src/ngAnimate/animateQueue.js:
```diff
@@ -624,25 +638,29 @@ var $$AnimateQueueProvider = ['$animateProvider', function($animateProvider) {
+          if (parentHost) {
+            // The pin target element becomes the parent element
+            parentElement = parentHost;
+            rootElementDetected = isMatchingElement(parentElement, $rootElement);
```
This change assumes that there can only be one "pinning" on the path from any element to the rootElement.
E.g. one can't pin element A to element B and element B to element C (where only C is a child of the rootElement).

Although this was not explicitly supported (nor do I know why one might want to do it), it is a potential BC.

gkalpak commented on src/ngAnimate/animateQueue.js:
```diff
@@ -624,25 +638,29 @@ var $$AnimateQueueProvider = ['$animateProvider', function($animateProvider) {
-              parentElement = parentHost;
-              rootElementDetected = true;
-            }
+        if (!rootElementDetected && !parentHost) {
```
This change assumes that there can only be one "pinning" on the path from any element to the rootElement.
E.g. one can't pin element A to element B and element B to element C (where only C is a child of the rootElement).

Although this was not explicitly supported (nor do I know why one might want to do it), it is a potential BC.

Narretz replied:
Good catch. I think we can add this as a reasonable breaking change. So far, pin() was broken for more cases than it was working for, so adoption doesn't seem to high.

Narretz committed:
perf(ngAnimate): speed up areAnimationsAllowed check

This commit speeds up the code that checks if an element can
be animated, for the following two cases:

The checks will be sped up in cases where the animation
is disabled via $animate.enabled(element, false) on any parent element.

A minor speed-up is also included for cases where the $rootElement of the
app (the bootstrap element) is on the body or lower in the DOM tree.

Narretz commented:
@gkalpak I've reverted the change, and recursive pinning is now allowed again. We can always change that later. And I've found and fixed another bug with the host element detection. Can you please review again?

gkalpak commented on test/ngAnimate/animateSpec.js:
```diff
@@ -1543,34 +1547,92 @@ describe("animations", function() {
+    they('should not animate an element when the pinned element (which is the $prop element), is pinned to an element that is not a child of the $rootElement',
+      ['same', 'parent', 'grandparent'],
+      function(elementRelation) {
+        inject(function($animate, $compile, $document, $rootElement, $rootScope) {
```
`$compile` is not used (here and below).

gkalpak commented on test/ngAnimate/animateSpec.js:
```diff
@@ -1543,34 +1547,92 @@ describe("animations", function() {
+    });
+
+    they('should adhere to the disabled state of the hosted parent when the $prop element is pinned',
+      ['parent'],
```
Why `they` if just using one value ?

Narretz commented:
@gkalpak Thanks for the review. Should be all good now.

gkalpak commented:
LGTM :smiley:
